.github/workflows/test-tpg.yml

```diff
@@ -14,36 +20,89 @@ on:
+        ref: ${{ github.event.inputs.branch }}
         path: provider
+        fetch-depth: 2
+    - name: Check for Code Changes
+      id: pull_request
+      run: |
+        cd provider
+        gofiles=$(git diff --name-only HEAD~1 | { grep -e "\.go$" -e "go.mod$" -e "go.sum$" || test $? = 1; })
+        if [ -z "$gofiles" ]; then
+          echo "has_changes=false" >> $GITHUB_OUTPUT
+        else
+          echo "has_changes=true" >> $GITHUB_OUTPUT
+        fi
+    - name: Get Job URL
+      if: ${{ !failure() && steps.pull_request.outputs.has_changes == 'true' }}
+      id: get_job
+      run: |
+        response=$(curl --get -Ss -H "Authorization: token ${{ secrets.GITHUB_TOKEN }}" -H "Accept: application/vnd.github.v3+json" "https://api.github.com/repos/${{ github.repository }}/actions/runs/${{ github.run_id }}/attempts/${{ github.run_attempt }}/jobs")
+        html_url=$(echo "$response" | jq -r --arg job_name "${{ github.job }}" '.jobs | map(select(.name == $job_name)) | .[0].html_url')
+        echo "url=${html_url}" >> $GITHUB_OUTPUT
+    - name: Post Pending Status to Pull Request
+      if: ${{ !failure() && steps.pull_request.outputs.has_changes == 'true' }}
+      run: |
+        curl -X POST -H "Authorization: token ${{secrets.GITHUB_TOKEN}}" \
+        -H "Accept: application/vnd.github.v3+json" \
+        "https://api.github.com/repos/GoogleCloudPlatform/magic-modules/statuses/${{github.event.inputs.sha}}" \
+        -d '{
+        "context": "${{ github.event.inputs.repo }}${{ env.status_suffix }}",
+        "target_url": "${{ steps.get_job.outputs.url }}",
+        "state": "pending"
+        }'
     - name: Set up Go
-      uses: golang/setup-go@v2
+      if: ${{ !failure() && steps.pull_request.outputs.has_changes == 'true' }}
+      uses: actions/setup-go@v4
       with:
-        go-version: '1.19.x'
-    - name: Build
+        go-version: '^1.19.1'
+    - name: Build Provider
+      if: ${{ !failure() && steps.pull_request.outputs.has_changes == 'true' }}
       run: |
         cd provider
         go build
-    - name: Lint
+    - name: Run Unit Tests
+      if: ${{ !failure() && steps.pull_request.outputs.has_changes == 'true' }}
+      run: |
+        cd provider
+        make test
+    - name: Lint Check
+      if: ${{ !cancelled() && steps.pull_request.outputs.has_changes == 'true' }}
       run: |
         cd provider
         make lint
-      if: ${{ always() }} # This step will always run, even if the previous fails
-    - name: Test
+    - name: Documentation Check
+      if: ${{ !cancelled() && steps.pull_request.outputs.has_changes == 'true' }}
       run: |
         cd provider
-        make test
-      if: ${{ always() }} # This step will always run, even if the previous fails
-
-
+        make docscheck
+    - name: Post Result Status to Pull Request
+      if: ${{ !cancelled() && steps.pull_request.outputs.has_changes == 'true' }}
```

Comment:
I think this should be `always()`, since we need to update this status even if the job is cancelled (in which case we should update the status to... I guess `failure` since `cancelled` isn't an option)

Reply:
hm, I'm not sure how to handle the case where the PR didn't cause changes (or was cancelled / errored before the changes). I would lean towards marking this as a "success" if it was triggered and ran to completion even if nothing changed. Otherwise there would be no way to know if it ran and failed.

I think the alternative would be to handle the check for has_changes in cloudbuild, and only run the github action if changes are detected. The action could still own its own status updates, and any issues during change detection would cause cloudbuild to fail.

Disadvantage being that then cloudbuild would need to check out the repository to check whether there are changes, which is currently removed entirely.

I don't know that I have a preference, but I think we need to make sure that this action _always_ reports a status when triggered, even if it is cancelled or has errors - otherwise there will be a risk of errors going undetected.

Reply:
backtracking on myself, but leaving previous comments in place for context - since we're auto-cancelling old action runs, `!cancelled()` is probably necessary. But the question of making sure that overall build failures are detected is still important. For example, if there's a github glitch and we're not able to clone the repo, I think that needs to show up as a failure for this job, either here or in cloudbuild, so that it can be rerun.

This was a bit of a brain dump - Let me know what you think!

Reply:
So concurrency conflict is going to be pretty rare here. Realistically I don't expect it to happen, but it's there if it happens. I debated omitted it entirely for simplicity. In the event of concurrent build we are assuming the SHA has changed. Since the `SHA` has changed `POST`ing to the old `SHA` is irrelevant. 

There is a argument that could be made in the case of a rerun against the same `SHA`. ie `/gcbrun`. In this case I think it is still correct to not post the status. Presumably the new job will `POST` the status so I don't want to collide with the old status. 

As for catching overall failures. The status of the builds should still be reflected in the `ACTIONS` view. If there is an issue on the PR the new job will post the status to the `SHA` so I think its best to omit collisions with the new jobs `POST` and just short circuit everything here.

Reply:
So, the issue I see is that because we're manually posting statuses, we need to be extra careful that the statuses are accurate.


If this action fails (or is cancelled!) before the final status post, the status will be "pending" indefinitely. For old SHAs, this would not be a critical issue per se but would likely be a little confusing to users. I would be fine with proceeding with the current behavior, though.

The real problem is that if this action fails before the _first_ status post, then it will _never show up_ on the PR as a status. (We're currently running into this as a problem related to TGC unit tests.) Since the action is tied to the main branch, not the PR, it also won't show up in the PR's actions tab. There won't be a way to rerun it or to even know that it should have run / started running.

One possible workaround would be to [require the status checks](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/managing-a-branch-protection-rule) which (IIRC) would tell GitHub to expect that they will eventually be there even if they haven't show up yet. (But in this case we would need to _always_ run those checks, not only if go was changed. But also, unit tests are fairly fast, so maybe it would be fine?)

Reply:
Looking over the PR again, I think you're already doing the status update as early as you can given the constraint of "only run these if there are changes." I'm happy to move forward with that for now. We can discuss changing that separately if it seems necessary in the future.

Reply:
To put it another way - this implementation doesn't have any issues the current implementation doesn't also have, and doesn't limit our options in the future.